File: src/teamlense_modules/agent/services/agent-auth.service.ts
import { env } from "../../../config/env";
import { comparePassword, randomToken, sha256 } from "../../../shared/auth/crypto";
import { signAgentToken } from "../../../shared/auth/jwt";
import { prismaClient } from "../../../shared/db/prisma-client";

const resolveAgentExpiry = (): Date => {
  const ttl = env.jwtAgentTtl.trim().toLowerCase();
  const now = Date.now();

  const match = ttl.match(/^(\d+)([smhd])$/);
  if (!match) {
    return new Date(now + 1000 * 60 * 60 * 24 * 30);
  }

  const value = Number(match[1]);
  const unit = match[2];

  const multiplier =
    unit === "s" ? 1000 :
    unit === "m" ? 1000 * 60 :
    unit === "h" ? 1000 * 60 * 60 :
    1000 * 60 * 60 * 24;

  return new Date(now + value * multiplier);
};

export class AgentAuthService {
  static async login(input: { email: string; password: string; deviceLabel?: string }) {
    const email = input.email.trim().toLowerCase();

    const user = await prismaClient.user.findUnique({
      where: { email },
      include: {
        organization: {
          select: {
            id: true,
            name: true,
            slug: true,
          },
        },
      },
    });

    if (!user) {
      throw new Error("Invalid email or password");
    }

    const validPassword = await comparePassword(input.password, user.passwordHash);
    if (!validPassword) {
      throw new Error("Invalid email or password");
    }

    if (user.status !== "ACTIVE") {
      throw new Error("User account is not active");
    }

    if (user.role !== "EMPLOYEE") {
      throw new Error("Desktop agent login is only available for employees");
    }

    const tokenId = randomToken(16);
    const agentToken = signAgentToken({
      userId: user.id,
      organizationId: user.organizationId,
      role: user.role,
      tokenId,
    });

    const expiresAt = resolveAgentExpiry();

    await prismaClient.agentToken.create({
      data: {
        organizationId: user.organizationId,
        userId: user.id,
        tokenHash: sha256(agentToken),
        label: input.deviceLabel?.trim() || "Desktop Agent",
        expiresAt,
      },
    });

    return {
      token: agentToken,
      expiresAt: expiresAt.toISOString(),
      user: {
        id: user.id,
        fullName: user.fullName,
        email: user.email,
        role: user.role,
      },
      organization: user.organization,
    };
  }
}

File: backend-go/internal/services/agent_auth_service.go
package services

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/teamlens/backend-go/internal/config"
	"github.com/teamlens/backend-go/internal/models"
)

type AgentAuthService struct {
	pool *pgxpool.Pool
	jwt  *JWTService
	cfg  *config.Config
}

func NewAgentAuthService(pool *pgxpool.Pool, jwt *JWTService, cfg *config.Config) *AgentAuthService {
	return &AgentAuthService{pool: pool, jwt: jwt, cfg: cfg}
}

func (s *AgentAuthService) Me(ctx context.Context, userID string) (*models.AgentLoginResponse, error) {
	var user struct {
		ID             string
		FullName       string
		Email          string
		Role           models.AuthRole
		Status         string
		OrganizationID string
		OrgName        string
		OrgSlug        string
	}

	err := s.pool.QueryRow(ctx,
		`SELECT u.id, u.full_name, u.email, u.role, u.status, u.organization_id,
		        o.name, o.slug
		 FROM users u
		 JOIN organizations o ON o.id = u.organization_id
		 WHERE u.id = $1`, userID,
	).Scan(
		&user.ID, &user.FullName, &user.Email, &user.Role, &user.Status,
		&user.OrganizationID, &user.OrgName, &user.OrgSlug,
	)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, errors.New("User not found")
		}
		return nil, fmt.Errorf("query user: %w", err)
	}

	if user.Status != "ACTIVE" {
		return nil, errors.New("User account is not active")
	}

	if user.Role != models.RoleEmployee {
		return nil, errors.New("Desktop agent login is only available for employees")
	}

	return &models.AgentLoginResponse{
		User: models.UserResponse{
			ID:       user.ID,
			FullName: user.FullName,
			Email:    user.Email,
			Role:     user.Role,
		},
		Organization: models.OrgResponse{
			ID:   user.OrganizationID,
			Name: user.OrgName,
			Slug: user.OrgSlug,
		},
	}, nil
}

func (s *AgentAuthService) Login(ctx context.Context, email, password string, deviceLabel *string) (*models.AgentLoginResponse, error) {
	email = strings.ToLower(strings.TrimSpace(email))

	var user struct {
		ID             string
		FullName       string
		Email          string
		PasswordHash   string
		Role           models.AuthRole
		Status         string
		OrganizationID string
		OrgName        string
		OrgSlug        string
	}

	err := s.pool.QueryRow(ctx,
		`SELECT u.id, u.full_name, u.email, u.password_hash, u.role, u.status, u.organization_id,
		        o.name, o.slug
		 FROM users u
		 JOIN organizations o ON o.id = u.organization_id
		 WHERE u.email = $1`, email,
	).Scan(
		&user.ID, &user.FullName, &user.Email, &user.PasswordHash,
		&user.Role, &user.Status, &user.OrganizationID,
		&user.OrgName, &user.OrgSlug,
	)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, errors.New("Invalid email or password")
		}
		return nil, fmt.Errorf("query user: %w", err)
	}

	if !ComparePassword(password, user.PasswordHash) {
		return nil, errors.New("Invalid email or password")
	}

	if user.Status != "ACTIVE" {
		return nil, errors.New("User account is not active")
	}

	if user.Role != models.RoleEmployee {
		return nil, errors.New("Desktop agent login is only available for employees")
	}

	tokenID := RandomToken(16)
	agentToken, err := s.jwt.SignAgentToken(user.ID, user.OrganizationID, user.Role, tokenID)
	if err != nil {
		return nil, err
	}

	expiresAt := time.Now().Add(s.cfg.JWTOAgentTTL)
	tokenHash := SHA256(agentToken)
	lbl := "Desktop Agent"
	if deviceLabel != nil && *deviceLabel != "" {
		lbl = *deviceLabel
	}

	_, err = s.pool.Exec(ctx,
		`INSERT INTO agent_tokens (id, organization_id, user_id, token_hash, label, status, expires_at, created_at)
		 VALUES ($1, $2, $3, $4, $5, 'ACTIVE', $6, NOW())`,
		RandomToken(16), user.OrganizationID, user.ID, tokenHash, lbl, expiresAt,
	)
	if err != nil {
		return nil, fmt.Errorf("save agent token: %w", err)
	}

	return &models.AgentLoginResponse{
		Token:     agentToken,
		ExpiresAt: expiresAt.Format(time.RFC3339),
		User: models.UserResponse{
			ID:       user.ID,
			FullName: user.FullName,
			Email:    user.Email,
			Role:     user.Role,
		},
		Organization: models.OrgResponse{
			ID:   user.OrganizationID,
			Name: user.OrgName,
			Slug: user.OrgSlug,
		},
	}, nil
}
